journal/src/main/scala/com/evolutiongaming/kafka/journal/util/ResourcePool.scala

```diff
@@ -0,0 +1,276 @@
+        state.objects.collectFirst {
+          case (id, obj) if obj.state.isFree => (id, obj)
+        }
+      }
+
+      def markObjectAsUsed(id: ObjectId, obj: AllocatedObject[F, O]): F[Unit] = {
+        stateRef.update { state =>
+          state.copy(objects = state.objects + (id -> obj.copy(state = AllocatedObject.State.Used)))
+        }
+      }
+
+      def allocateObject(state: State[F, O]): F[(ObjectId, O)] = {
+        for {
+          a <- resource.allocated
+          (obj, release) = a
+          objId <- Sync[F].delay(UUID.randomUUID())
+          state1 = state.copy(
+            objects = state.objects + (objId -> AllocatedObject(obj, release, AllocatedObject.State.Used))
+          )
+          _ <- stateRef.set(state1)
+        } yield (objId, obj)
+      }
+
+      def checkForAcquireTimeout(
+        waiterId: WaiterId,
+        promise: Waiter[F, O]
+      ): F[Unit] = {
+        promise.tryGet.flatMap {
+          case Some(_) =>
+            Async[F].unit
+          case None =>
+            mutex.lock.surround {
+              promise.tryGet.flatMap {
+                case Some(_) =>
+                  Async[F].unit
+                case None =>
+                  for {
+                    _ <- promise.complete(
+                      Left(
+                        new TimeoutException(s"Resource acquisition timed out")
+                      )
+                    )
+                    _ <- stateRef.update { state =>
+                      state.copy(waiters = state.waiters - waiterId)
+                    }
+                  } yield ()
+              }
+            }
+        }
+      }
+
+      def registerWaiter(state: State[F, O], promise: Waiter[F, O]): F[Unit] = {
+        for {
+          waiterId <- Sync[F].delay(UUID.randomUUID())
+          _ <- stateRef.set {
+            state.copy(
+              waiters = state.waiters + (waiterId -> promise)
+            )
+          }
+
+          _ <- (Async[F].sleep(acquireTimeout) *> checkForAcquireTimeout(waiterId, promise)).start
+        } yield ()
+      }
+
+      val acquirePromise: F[Waiter[F, O]] =
+        Deferred[F, Either[Throwable, (ObjectId, O)]].flatMap { promise =>
+          mutex.lock.surround {
+            stateRef.get.flatMap { state =>
+              findFreeObject(state) match {
+                case Some((id, obj)) =>
+                  for {
+                    _ <- markObjectAsUsed(id, obj)
+                    _ <- promise.complete((id, obj.value).asRight)
+                  } yield ()
+
+                case None =>
+                  if (state.objects.size < poolSize) {
+                    for {
+                      o <- allocateObject(state)
+                      _ <- promise.complete(o.asRight)
+                    } yield ()
+                  } else {
+                    registerWaiter(state, promise)
+                  }
+              }
+            }
+          }.as(promise)
```

Comment:
I think you can also use `flatTap`, if you want to avoid `as(promise)` like this:
```scala
Deferred[F, Either[Throwable, (ObjectId, O)]].flatTap { promise =>
  ... your code ...
}
```

Reply:
done